packages/ui/src/components/text.tsx

```diff
@@ -1,138 +1,184 @@
-
-  /**
-   * Sets the color property.
-   */
-  color?: 'primary' | 'secondary' | 'tertiary' | 'tertiaryBackground' | 'foreground-5' | 'foreground-4' | 'inherit'
-
-  /**
-   * Controls the wrapping behavior of the text.
-   */
-  wrap?: 'wrap' | 'nowrap' | 'pretty' | 'balance'
+type TextVariant = Exclude<VariantProps<typeof textVariants>['variant'], undefined | null>
+
+const textVariantToElement: Record<TextVariant, TextElement> = {
+  'heading-hero': 'p',
+  'heading-section': 'p',
+  'heading-subsection': 'p',
+  'heading-base': 'p',
+  'heading-small': 'p',
+  'body-normal': 'p',
+  'body-single-line-normal': 'p',
+  'body-strong': 'p',
+  'body-single-line-strong': 'p',
+  'body-code': 'pre',
+  'caption-normal': 'span',
+  'caption-soft': 'span',
+  'caption-single-line-normal': 'span',
+  'caption-single-line-soft': 'span'
 }
 
-const Text = ({
-  className,
-  children,
-  as,
-  asChild,
-  align,
-  size = 2,
-  weight = 'normal',
-  color = 'primary',
-  trim,
-  truncate,
-  wrap,
-  ...props
-}: TextProps) => {
-  const Comp = asChild ? Slot : ((as ? as.toString() : 'span') as any)
-
-  return (
-    <Comp className={cn(textVariants({ align, size, weight, color, trim, truncate, wrap }), className)} {...props}>
-      {children}
-    </Comp>
-  )
+const getTextNode = ({ as, variant = 'body-normal', asChild }: Pick<TextProps, 'as' | 'asChild' | 'variant'>) => {
+  if (asChild) return Slot
+
+  if (as) return as
+
+  if (textVariantToElement[variant ?? 'body-normal']) {
+    return textVariantToElement[variant ?? 'body-normal']
+  }
+
+  return 'span' as ElementType
 }
 
-Text.displayName = 'Text'
+type TextProps<E extends TextElement = 'span'> = Omit<ComponentProps<E>, 'color'> &
+  VariantProps<typeof textVariants> & {
+    /**
+     * Shorthand for changing the default rendered element
+     * into a semantically appropriate alternative.
+     */
+    as?: TextElement
+    /**
+     * Change the default rendered element for the one
+     * passed as a child, merging their props and behavior.
+     */
+    asChild?: boolean
+  }
+
+type TextComponent = <E extends TextElement = 'span'>(
+  props: TextProps<E> & { ref?: Ref<HTMLElement> }
+) => ReactElement | null
+
+const TextWithRef = forwardRef<HTMLElement, TextProps>(
+  (
+    {
+      className,
+      children,
+      truncate = false,
+      variant = 'body-normal',
+      asChild,
+      align = 'left',
+      color = 'foreground-2',
+      wrap,
+      as,
+      ...props
+    },
+    ref
+  ) => {
+    const [withTooltip, setWithTooltip] = useState(false)
+    const localRef = useRef<HTMLElement>(null)
+
+    const Comp = getTextNode({ as, variant, asChild })
+
+    const isHeading = !as && !!variant?.startsWith('heading')
+
+    const compRef = useMergeRefs<HTMLElement>([localRef, ref])
+
+    useLayoutEffect(() => {
+      if (!truncate) return
+      if (!localRef.current) return
+
+      const { scrollWidth, scrollHeight, clientWidth, clientHeight } = localRef.current
```

Comment:
all four of these properties trigger a reflow in the browser if the layout is "dirty". hence the above comment.

Reply:
Yeah, sounds like a better approach. If `truncate` is `true`, then add `title` as the `innerText` of the child.